### PR TITLE
STAC-16579 add agent stateful check class

### PR DESCRIPTION
**Pull request by hruhek**
### Step 1: Link to Jira issue

* https://stackstate.atlassian.net/browse/STAC-16579
* https://stackstate.atlassian.net/browse/STAC-16108

### Step 2: Description of changes
Adding agent stateful check class. Check Jira story for details.
Added @MelcomVE changes from `stac-16108-transactional-c-binding` branch.

### Step 3: Did you add / update tests for your changes in the right area?
- [x] Unit/Component test
- [ ] Integration test	

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    """
+    Agent error that needs to be converted to Critical Service Check
+    """
+
+    def to_string(self):
+        # type: () -> str
+        return "check error: {}".format(self.message)
+
+
+class CheckMixin(object):
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:38]: Maybe this is CheckSetupMixin? If I understand this mixin is used only to declare the `setup` method, right?

File: stackstate_checks_base/stackstate_checks/base/checks/__init__.py [hunk 2]
```diff
@@ -21,5 +22,10 @@
     'EventType',
     'HealthStream',
     'HealthStreamUrn',
-    'Health'
+    'Health',
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/__init__.py:25]: If I understand correctly, the __all__ is used to make these classes public when importing this module.
Is there a reason on why we would need to make `Health`, `Stateful` and `Transactional` public classes?

File: stackstate_checks_base/stackstate_checks/base/utils/transactional_api.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    import transaction
+
+    using_stub_transaction = False
+except ImportError:
+    from ..stubs import transaction
+
+    using_stub_transaction = True
+
+
+class TransactionApi(object):
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/utils/transactional_api.py:19]: Would it make more sense if we call it `TransactionManager`?

File: stackstate_checks_base/stackstate_checks/base/utils/state_api.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    import state
+
+    using_stub_state = False
+except ImportError:
+    from ..stubs import state
+
+    using_stub_state = True
+
+
+class StateApi(object):
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/utils/state_api.py:19]: Same as Transaction, maybe StateManager?

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
+import copy
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:1]: Maybe move the new V2 checks to `base/checks/v2/base.py`. Also, the `CheckMixin`, `Health` and `Stateful` could be in an `util/v2` module. Whoever is using the new base class doesn't need to import them.

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+
+        self.set_state_transactional(new_state)
+        self.transaction.stop()
+
+        return
+
+    def transactional_check(self, instance, state):
+        # type: (_InstanceType, Union[Dict[str, Any], Model]) -> (Dict[str, Any], Optional[CheckError])
+        """
+        This method should be implemented for a Stateful Check. It's called from run method.
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:323]:
```suggestion
        This method should be implemented for a Transactional Check. It's called from run method.
```

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+        All Errors raised from stateful_check will be caught and converted to service_call in the run method.
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:324]:
```suggestion
        All Errors raised from transactional_check will be caught and converted to service_call in the run method.
```

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+                older versions of the Agent).
+        """
+        # Initialize AgentCheck's base class
+        super(Transactional, self).__init__(*args, **kwargs)
+        self.transaction = None  # type: Optional[TransactionApi]
+
+    def set_state_transactional(self, new_state):
+        # type: (Union[Dict, Model]) -> None
+        """
+        Set new checks state.
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:167]:
```suggestion
        Set new checks state in the transaction.
```

**Review by julianosk — COMMENTED** [review]
It looks good overall. Only a few details. (I haven't gone through everything)

File: stackstate_checks_base/stackstate_checks/base/checks/base_v2.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+        return self.state.get(self.PERSISTENT_CACHE_KEY)
+
+    def setup(self):
+        # type: () -> None
+        if self.state is not None:
+            return None
+        self.state = StateApi(self)
+
+
+class Transactional(Stateful):
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/base_v2.py:144]: Could it be `TransactionalBase`?

File: stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    def stateful_check(self, instance, state):
+        # type: (InstanceType, StateType) -> (StateType, Optional[CheckError])
+        """
+        This method should be implemented for a Stateful Check. It's called from run method.
+        All Errors raised from stateful_check will be caught and converted to service_call in the run method.
+
+        This method should be implemented for a Stateful Check. It's called from run method.
+        All Errors raised from stateful_check will be caught and converted to service_call in the run method.
+
```
Comment (hruhek) [inline stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py:42]: This is duplicate line as 37-39. Please delete.

File: stackstate_checks_base/stackstate_checks/base/checks/v2/types.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+InstanceType = TypeVar('InstanceType', Model, Dict[str, Any])
+StateType = TypeVar('StateType', Model, Dict[str, Any])
+
+
+class CheckResponse(object):
+    """
+    CheckResponse is the return type of the check function of Agent V2 checks
+    """
+    def __init__(self, *args, **kwargs):
```
Comment (julianosk) [inline stackstate_checks_base/stackstate_checks/base/checks/v2/types.py:12]: A learning question: what is the advantage of having `*args`, `**kwargs` instead of named arguments with `None` as default? Like:
```python
def __init__(self, persistent_state = None, transactional_state = None, check_error = None):
```

File: stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    def stateful_check(self, instance, persistent_state):
+        # type: (InstanceType, StateType) -> CheckResponse
+        """
+        This method should be implemented for a Stateful Check. It's called from run method.
+        All Errors raised from stateful_check will be caught and converted to service_call in the run method.
+
+        This method should be implemented for a Stateful Check. It's called from run method.
+        All Errors raised from stateful_check will be caught and converted to service_call in the run method.
+
```
Comment (hruhek) [inline stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py:37]: Duplicate line, please delete

File: stackstate_checks_base/tests/test_agent_check_v2.py [hunk 13]
```diff
@@ -0,0 +1,1610 @@
...
+TEST_INSTANCE = {
+    "url": "https://example.org/api"
+}
+
+
+def test_instance():
+    """
+    Simply assert the class can be instantiated
+    """
+    AgentCheckV2()
```
Comment (hruhek) [inline stackstate_checks_base/tests/test_agent_check_v2.py:33]: Why is this here? Its not assigned to variable.

**Review by hruhek — COMMENTED** [review]
I created this PR so GitHub doesn't allow me to request changes to myself. 
Please check the comments @zandrewitte.

File: stackstate_checks_base/tests/test_agent_check_v2.py (hunk 13 above)
Comment (zandrewitte) [reply stackstate_checks_base/tests/test_agent_check_v2.py:33]: No idea, was inherited from `test_agent_check` so I'll remove it 😄

File: stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py (hunk 12 above)
Comment (zandrewitte) [reply stackstate_checks_base/stackstate_checks/base/checks/v2/stateful_agent_check.py:37]: Fixed

**Review by hruhek — COMMENTED** [review]
Looks good to me :)
